progress on api

jeffreyhanson opened the pull request:
Hi,

I've been coding up some ideas/scaffolding for the API. @dimitri-justeau, when you get a chance, could you plese take a look? It still needs work before its ready to be merged - but I'd just like to get your input to make sure I'm not going off on a totally wrong directon? Based on our discussion, I was thinking that ths API could ultimately look something like this:

```
solution <- 
  rstopt_problem(habitat_data, restorable_data) %>%
  add_min_icc_objective() %>%
  add_components_constraints(min = 1, max = 3) %>%
  solve(number_solutions = 100, time_limit = 20)
```

jeffreyhanson committed:
progress on api

jeffreyhanson commented:
Also, I've got other stuff I need to focus on today, so please feel free to work on this branch.

dimitri-justeau commented on R/restopt_problem.R:
```diff
@@ -0,0 +1,87 @@
+    )
+  )
+
+  # return object
```
If I understand correctly, the initial S4 class would become a simpler data frame, that would be processed in the `restopt_compile` function? I'm ok to go with that, as it seems simple and efficient!

jeffreyhanson replied:
Yeah, that's what I was thinking. The `restopt_problem()` function would create an S3 class object (basically a list with a defined class) that could be updated using objective/constraint functions. After the user has finished specifying their problem, they would call `solve()` to (1) build the problem specification through the Java API (via the `compile()` function), (2) run the optimization, and (3) return optimization results (e.g. as a `terra::rast()` object).

jeffreyhanson replied:
Ah - I just noticed you described these exact steps in a later comment

jeffreyhanson replied:
I think we're in agrement, so I'll mark this comment thread as resolved - but please feel free to re-open if you want to chat more about this.

dimitri-justeau commented on R/solve.R:
```diff
@@ -0,0 +1,31 @@
+#' @examples
+#' # TODO
+#'
+#' @export
```
Ok so that would be the generic solving function, with the four basic steps: compile, solve using java, prepare the results and return them. It seems perfect, I just wonder why calling it `solve.RestoptProblem`, instead of using something like `restoptr::solve(problem`, it is an R naming convention ?

jeffreyhanson replied:
Yeah - that's I was thinking. Awesome! Yeah, method dispatch is a bit more compliated when there are functions with the same name already present in base R packages (or in other commonly used R packages). Since the base R package already contains a `solve()` function, we can't just assign a `solve` function here because it would clash - we need to implement the function in a way that plays nicely with base R. In particular, base R solve function (https://www.rdocumentation.org/packages/base/versions/3.6.2/topics/solve) provides an S3 generic. This means that if we create a `solve.RestoptProblem` function then R will automatically use this `solve.RestoptProblem` function if the user runs `solve(x)` where `x` is a `RestoptProblem` class object. In other words, S3 methods are a way to dispatch functions/methods using a `${method_name}.${object_class}` style suffix system based on the first argument/parameter (S4 methods are handy because they let you dispatch on multiple parameters/arguments). Does that make sense? Sorry, I'm probably getting some of this terminology wrong. If it's helpful, here's some info on S3 objects and methods: http://adv-r.had.co.nz/S3.html

dimitri-justeau replied:
It is perfectly clear :-)

jeffreyhanson replied:
Awesome!

dimitri-justeau commented on R/restopt_problem.R:
```diff
@@ -0,0 +1,87 @@
+#' values that indicate if each planning unit contains habitat or not.
+#'
+#' @param restorable_habitat [terra::rast()] Raster object containing
+#' continuous values that indicate the amount of habitat that can
```
Just a detail (I can correct it later): restorable values must be integers.

jeffreyhanson replied:
Ah - thanks for pointing that out. On that note, maybe it would be worth updating the argument checks to verify that the raster data values are sensible? E.g., no negative values, all values range between sensible values, rasters have at least one non-missing values? We have these checks in the prioritizr package, but they can be slow to run for very large datasets. What do you think?

dimitri-justeau replied:
Yes, it would be really nice to have such checks. Maybe we can list these in a specific issue, or keep them in mind to add them once we have the base functionalities of the package working?

jeffreyhanson replied:
I've started adding some checks to `restopt_problem` - is there a sensible upper limit for `restorable_habitat`?

dimitri-justeau commented on R/restopt_component.R:
```diff
@@ -0,0 +1,30 @@
+#' @examples
+#' # TODO
+#'
+#' @export
```
So this is the generic function to create a component (constraint, objective, ...), and it contains the name of the corresponding java method to call, and its arguments, am I right?

dimitri-justeau replied:
Also, I am wondering, is it really useful to have such a file, as it will never be used, and it does not factorize any code?

jeffreyhanson replied:
Yeah that's right. I was thinking that this class/function/constructor could be useful to store data/information for each constraint/objective in a standard way. After sotring the data/information for an arbitrary constraint/objective in a standard way, I was thinking that the `solve()` function could read the information/data stored in these objects and use them to build the problem specification using the Java API. I was also thinking another advantage of this class/function/constructor is that it could be run data sanity checks (e.g. ensure that the `java` parameter contains a valid `character` value) to minimize chances of passing invalid values to the Java API. Does that make sense? I completely agree that we want to avoid bloating the code - maybe there's a better way to do this?

dimitri-justeau replied:
Actually, this function is useful and do factorize code, I was mistaken in the first place.

jeffreyhanson replied:
Ok great - I think we're in agrement, so I'll mark this comment thread as resolved

dimitri-justeau reviewed (commented):
Overall it feels very nice to organize things like that, it is not too complicated but at the same time provides flexibility. I approve! If you are ok, we can work together on this branch.

dimitri-justeau commented:
> Also, I've got other stuff I need to focus on today, so please feel free to work on this branch.

No problem, I also have other things to do at the same time. I had a look at this branch and I like how you organized thing, thank you a lot !

dimitri-justeau commented on R/add_compactness_constraints.R:
```diff
@@ -0,0 +1,53 @@
+#' @include internal>R
+NULL
+
+#' Add constraints to limit compactness
```
The compactness constraint is a single constraint, I would have named it `add_compactness_constraint`, but I am not a native English speaker, so I might also be missing something here.

jeffreyhanson replied:
Ah yeah, that's a good point. Personally, I think the important thing here is consistency - so that all functions for adding a constraint(s) (e.g. `add_compactness_constraint`, `add_components_constraint`, `add_locked_out_constraint`, etc) have the same suffix to indicate that they belong to the same family of functions. I worry that having some of functions ending in `constraints` and others in `constraint` might potentially be confusing for users?  So for example, we could have a `_constraint` suffix, `_constraints` suffix, or something else? How does that sound?

dimitri-justeau replied:
I agree, and I would go with the `_constraint` suffix.

jeffreyhanson replied:
Ok - sounds good - let's go with that.

dimitri-justeau commented on R/restopt_component.R:
```diff
@@ -0,0 +1,30 @@
+#' @param data `list` Object containing parameters for adding the component.
+#'  Defaults to an empty list.
+#'
+#' @param args `character` Vector used to map the names of the elements in
```
This won't help, as arguments in Java methods are not named. I'd rather avoid trying to design a generic mapping with the Java API, as I think it will just overcomplicate things in our case (Java calls will only consist of one line per constraint/objective, and I do not think we will exceed 10 constraints).

jeffreyhanson replied:
Ah - thank you very much for catching! You're right, I mistakenly assumed we could use named arguments to map with the Java API. So does that mean that as long as the arguments are specified in the right order, then it should work? If so, then we could just make sure that the elements in `data` are in the right order for the Java code - and that should work? Is that right?

dimitri-justeau replied:
Yes in Java methods/functions arguments are ordered, we just need to call the corresponding java function in the correct order.

jeffreyhanson replied:
Ah ok - excellent!

dimitri-justeau reviewed (commented):
I had a closer look, I'd like to simplify some things, notably the link with Java which does not need to be generic. Communication with Java will only consist of a few lines, and will not grow much in the future. For instance, the locked-out constraint actually defines how the problem will be constructed in Java and is not posted as the other constraints in Choco. It is convenient to define it as a constraint for the user, but in restopt internal logic, it is not really a constraint. As a consequence, the corresponding Java operations are different from the compactness constraint for instance.

dimitri-justeau committed:
Start adapting API proposed in PR https://github.com/dimitri-justeau/restoptr/pull/7 according to comments

dimitri-justeau committed:
Start adapting API proposed in PR https://github.com/dimitri-justeau/restoptr/pull/7 according to comments

dimitri-justeau commented:
I adapted a few thing (still in progress) from your proposed API and according to my comments in this branch : https://github.com/dimitri-justeau/restoptr/tree/api_dim

jeffreyhanson commented:
Awesome - thanks! I really like your suggestion for simplifying the component object by using `post` field that has a function. I think we could further simplify it by removing the `data` field (since the `post` function will inherit objects from the environment where it was defined). I'm hoping I'll have time to this afteroon to play around with this. So, I'll merge your `api_dim` branch into the `api` branch - and then keep adding commits to `api` branch.

jeffreyhanson committed:
Merge pull request #8 from dimitri-justeau/api_dim

merge Dimitri's edits

jeffreyhanson committed:
add stuff for restopt_problem
- add a settings element to store time limit and precision
- add add_restopt_objective() to standardize methods for adding objective
- add add_restopt_constraint() to standardize methods for adding constraint

jeffreyhanson committed:
add stuff for restopt_problem
- add a settings element to store time limit and precision
- add add_restopt_objective() to standardize methods for adding objective
- add add_restopt_constraint() to standardize methods for adding constraint

jeffreyhanson committed:
update restopt_component function
- remove uneeded data slot
- add documentation for post parameter

jeffreyhanson committed:
update restopt_component function
- remove uneeded data slot
- add documentation for post parameter

jeffreyhanson committed:
update add_*_constraint functions
- these now use standard method for adding constraints

jeffreyhanson committed:
update add_*_constraint functions
- these now use standard method for adding constraints

jeffreyhanson committed:
update add_*_objective functions
- these now use standard method for adding objs
- attempt to add post method for adding objs and generating solutions

jeffreyhanson committed:
update add_*_objective functions
- these now use standard method for adding objs
- attempt to add post method for adding objs and generating solutions

jeffreyhanson committed:
add fns for handling disk io for terra

jeffreyhanson committed:
add fns for handling disk io for terra

jeffreyhanson committed:
update restorable constraint to use standard method

jeffreyhanson committed:
update restorable constraint to use standard method

jeffreyhanson committed:
draft code for solve() function
- also remove compile function
- this is because the restopt api will generating a solution once we specify the obj, so can't build and solve the problem in seperate steps

jeffreyhanson committed:
draft code for solve() function
- also remove compile function
- this is because the restopt api will generating a solution once we specify the obj, so can't build and solve the problem in seperate steps

jeffreyhanson committed:
update docs

jeffreyhanson committed:
update docs

jeffreyhanson committed:
add method for specifying settings

jeffreyhanson committed:
add method for specifying settings

jeffreyhanson commented:
I've added some suggestions for streamlining the code. @dimitri-justeau, when you get a chance, could you please take a look? I've tried to include explanations of the rationale behind the changes in the commit messages - but please let me know if anything's not clear?

Also, I was wondering if you had any ideas on the best way to handle the accessibility data when initializing the problem (i.e. [here](https://github.com/dimitri-justeau/restoptr/blob/api/R/solve.R#L32-L41)). From what I can tell, the accessibility data needs to be added when initializating the problem, so do we need to have some code that specifically checks `x$constraints` for locked out constraints and then (i) if the constraint is present, add it to the data to the loader and use it to create the problem; or (ii) if the constraint is not present, create a default dataset (with every restorable or existing habitat cell set to 1) and use that instead? Or is there a more elegant way to handle this?

jeffreyhanson committed:
remove unused ... post parameters

jeffreyhanson committed:
remove unused ... post parameters

jeffreyhanson commented:
Sorry - just realized that there were some unused arguments in the post functions - should be good now. I'll hold off on making any further changes until you've had a chance to look at it.

dimitri-justeau committed:
Update API: \n - Fix some bugs / typos \n - Complete previous work (e.g. use standard method in add_locked_out_constraint) \n - Use explicit argument name to facilitate code comprehension

dimitri-justeau committed:
Update doc

dimitri-justeau commented:
Ok, I fixed some typos in the API, and I also suggest using explicit argument names instead of `x` or `y` for a more readable code, you can see my changes in this commit: https://github.com/dimitri-justeau/restoptr/commit/ec741c4746a6d47fa93bc0f0f93417b4a0f27624

jeffreyhanson commented:
Sorry for my slow response. Yeah, that's awesome - thank you very much! I'll take another look at it this weekend.

Also, just in case this is helpful, I find the tidyverse style guide useful keeping code consistent (see https://style.tidyverse.org/index.html) (but I do use camelcase for classes - which I guess from other languages). If you're using Rstudio (or Atom - that's what I use), I think it's possible to set up a linter to automatically scan files as you work on them for style consistency. Just an idea - please feel free to ignore if you don't think it's worth the effort.

dimitri-justeau committed:
Start fixing solve.R

jeffreyhanson committed:
Merge branch 'api' into api_dim

jeffreyhanson committed:
Merge pull request #9 from dimitri-justeau/api_dim

Merge api_dim branch

jeffreyhanson commented:
@dimitri-justeau, just FYI, I've merged the additional commits from the api-dim branch in the api branch.

jeffreyhanson committed:
remake docs

jeffreyhanson committed:
fix doc issue raised by check_man()

jeffreyhanson commented:
@dimitri-justeau, thanks for updating all the functions to have sensible/informative parameter names! This is great! I just wanted to note that we need to follow certain conventions for the `print.RestoptProblem` and `solve/RestoptProblem` S3 methods in order to pass CRAN checks. When I run CRAN package checks on my computer (using `devtools::check()`), I see this:

```
❯ checking S3 generic/method consistency ... WARNING
  print:
    function(x, ...)
  print.RestoptProblem:
    function(problem, ...)
  
  solve:
    function(a, b, ...)
  solve.RestoptProblem:
    function(problem, ...)
  
  See section ‘Generic functions and methods’ in the ‘Writing R
  Extensions’ manual.
```

This is because the base R package provides S3 generics for the `print` and `solve` functions. To pass the CRAN checks, we need to use the same parameters for our `print.RestoptProblem` and `solve.RestoptProblem` S3 methods as the S3 generics (i.e. `print` and `solve` above). This is important to make sure that the package plays nicely with other R packages. So, I'll push a commit that updates the parameters for `print.RestoptProblem` and `solve.RestoptProblem` so they pass CRAN checks. 

Does that make sense? Sorry if I'm not explaining this very well - happy to chat over zoom/skype to explain this further if that would be helpful?

jeffreyhanson committed:
fix print and solve S3 method consistency

jeffreyhanson committed:
lint code
i.e. ensure line length <= 80 characters

jeffreyhanson committed:
add default locked out values
this is needed because the solve() function assumes locked out data are provided

jeffreyhanson committed:
only set default Java parameters if none specified
this is useful so that users can set >4 GB memory if they want

jeffreyhanson committed:
update onLoad fn
- put restopt jar download URL in variable to simplify updates for future versions
- print package start up message if downloading file to explain why a file is being downloaded
- save jar file in directory  that (i) is CRAN approved and (ii) users have write permissions (users might not be able to write into pkg dir if pkg installed by HPC admin)

jeffreyhanson committed:
remove unused jar files
- jar files now stored in tools user dir

jeffreyhanson committed:
add in example data files from main branch
- these files got deleted some how?

jeffreyhanson committed:
export pipe operator

jeffreyhanson committed:
add magrittr to pkg imoprts for pipe

jeffreyhanson committed:
add more pipe stuff

jeffreyhanson committed:
tweak restoptr print
- use message() instead of cat(), because this can easily be suppressed if needed
- show parameter values for constraints

jeffreyhanson committed:
revert tools dir for java
- couldnt get working

jeffreyhanson committed:
revert deletion of java files

jeffreyhanson committed:
add checks for raster cell values

jeffreyhanson commented:
@dimitri-justeau, I've added some more tweaks. Could you please take a look? In particular, please let me know if you have any ideas for making the `print.RestoptProblem` method look nicer?

dimitri-justeau commented:
> @dimitri-justeau, thanks for updating all the functions to have sensible/informative parameter names! This is great! I just wanted to note that we need to follow certain conventions for the `print.RestoptProblem` and `solve/RestoptProblem` S3 methods in order to pass CRAN checks. When I run CRAN package checks on my computer (using `devtools::check()`), I see this:
> 
> ```
> ❯ checking S3 generic/method consistency ... WARNING
>   print:
>     function(x, ...)
>   print.RestoptProblem:
>     function(problem, ...)
>   
>   solve:
>     function(a, b, ...)
>   solve.RestoptProblem:
>     function(problem, ...)
>   
>   See section ‘Generic functions and methods’ in the ‘Writing R
>   Extensions’ manual.
> ```
> 
> This is because the base R package provides S3 generics for the `print` and `solve` functions. To pass the CRAN checks, we need to use the same parameters for our `print.RestoptProblem` and `solve.RestoptProblem` S3 methods as the S3 generics (i.e. `print` and `solve` above). This is important to make sure that the package plays nicely with other R packages. So, I'll push a commit that updates the parameters for `print.RestoptProblem` and `solve.RestoptProblem` so they pass CRAN checks.
> 
> Does that make sense? Sorry if I'm not explaining this very well - happy to chat over zoom/skype to explain this further if that would be helpful?

Yes it makes total sense, thank you for pointing this out !

dimitri-justeau commented:
> @dimitri-justeau, I've added some more tweaks. Could you please take a look? In particular, please let me know if you have any ideas for making the `print.RestoptProblem` method look nicer?

Thank you! I did not have much time to focus on this in the last few days, I'll have a look today. Then I will be on holiday from 21 December to 3 January, so don't worry if I don't answer during this period (if you are not in holiday too during this period). Thank you again for all your help !

dimitri-justeau commented:
I corrected a few things in my api_dim branch. Tests are passing, it seems like the core ideas of the package are settled. After holidays, I believe that the next steps are:

- Implement what you suggested in issue https://github.com/dimitri-justeau/restoptr/issues/4
- Complete documentation and examples
- See how we can make the print.RestoptProblem method look nicer

I think that once we'll have that, we will be able to make the first version of `restoptr` (0.1.0). After that, I'd like to focus on automatic input preprocessing, as explained in my last comment in issue https://github.com/dimitri-justeau/restoptr/issues/10

@jeffreyhanson, if you are still available and interested to collaborate, we could organize a Skype meeting in January to share ideas and plan the next steps?

Cheers !

jeffreyhanson commented:
Awesome - thank you very much@dimitri-justeau! I'll take a closer a look when I can. Yeah, definitely, let's organize a skype meeting in January, I'll send you an email.
